pypit/par/pypitpar.py

```diff
@@ -0,0 +1,2486 @@
+        k = cfg.keys()
+        parkeys = [ 'method', 'params' ]
+        kwargs = {}
+        for pk in parkeys:
+            kwargs[pk] = cfg[pk] if pk in k else None
+        return cls(**kwargs)
+
+    @staticmethod
+    def valid_methods():
+        """
+        Return the valid overscane methods.
+        """
+        return [ 'polynomial', 'savgol', 'median' ]
+
+    def validate(self):
+        """
+        Check the parameters are valid for the provided method.
+        """
+        if self.data['method'] is None:
+            return
+        if self.data['params'] is None:
+            raise ValueError('No overscan method parameters defined!')
+
+        if self.data['method'] == 'polynomial' and len(self.data['params']) != 3:
+            raise ValueError('For polynomial overscan method, set params = order, number of '
+                             'pixels, number of repeats')
+
+        if self.data['method'] == 'savgol' and len(self.data['params']) != 2:
+            raise ValueError('For savgol overscan method, set params = order, window size')
+            
+        if self.data['method'] == 'median' and self.data['params'] is not None:
+            warnings.warn('No parameters necessary for median overscan method.  Ignoring input.')
+
+
+class FlatFieldPar(ParSet):
+    def __init__(self, frame=None, method=None, params=None, twodpca=None):
+    
+        # Grab the parameter names and values from the function
+        # arguments
+        args, _, _, values = inspect.getargvalues(inspect.currentframe())
+        pars = dict([(k,values[k]) for k in args[1:]])
+
+        # Initialize the other used specifications for this parameter
+        # set
+        defaults = dict.fromkeys(pars.keys())
+        options = dict.fromkeys(pars.keys())
+        dtypes = dict.fromkeys(pars.keys())
+        descr = dict.fromkeys(pars.keys())
+
+        # Fill out parameter specifications.  Only the values that are
+        # *not* None (i.e., the ones that are defined) need to be set
+
+        # TODO: Provide a list of valid masters to use as options?
+        defaults['frame'] = 'pixelflat'
+        dtypes['frame'] = str
+        descr['frame'] = 'Frame to use for field flattening.  Options are: pixelflat, pinhole, ' \
+                         'or a specified master calibration file.'
+
+        defaults['method'] = 'bspline'
+        options['method'] = FlatFieldPar.valid_methods()
+        dtypes['method'] = str
+        descr['method'] = 'Method used to flat field the data; use None to skip flat-fielding.  ' \
+                          'Options are: None, {0}'.format(', '.join(options['method']))
+
+        defaults['params'] = 20
+        dtypes['params'] = [int, list]
+        descr['params'] = 'Flat-field method parameters.  For \'PolyScan\', set params = order, ' \
+                          'numPixels, repeat ; for bspline, set params = spacing '
+
+        # TODO:  How is twodpca used?  Is it just another method that is
```

Comment:
Good question for @rcooke-ast

Reply:
Okay, I'll leave it for now, and then we can simultaneously remove the keyword and the functionality later.